PhysicsExploration/RenderingPipeline.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include "CallbackHandler.h"
+#include "Model.h"
+#include "Camera.h"
+#include "Lighting.h"
+#include "MenuGUI.h"
+#include "stb_image.h"
+#include "Shader.h"
+#include <fstream>
+#include <string>
+#include <functional>
+class RenderingPipeline : CallbackHandler
+{
+public:
+	RenderingPipeline();
+	~RenderingPipeline();
+	int runAppTemplate();
+	void processInput() const;
+protected:
+	int initializeWindow() const;
+	void initializeCallbacks() const;
+	virtual void initializeShaders() const = 0;
+	virtual void initializeModels() const = 0;
+	void static configureDepthBuffer();
+	void addLights() const;
+	void processFrame() const;
+	void firstPass() const;
+	void secondPass() const;
+	virtual void renderScene(const Shader& shader) const = 0;
+	virtual void framebuffer_size_callback(GLFWwindow* window, int width, int height);
+	virtual void key_callback(GLFWwindow* window, int key, int scancode, int action, int mod);
+	virtual void cursor_pos_callback(GLFWwindow* window, double xpos, double ypos);
+	virtual void scroll_callback(GLFWwindow* window, double xoffset, double yoffset);
+
+	// Window options
+	const GLuint WIDTH = 800;
+	const GLuint HEIGHT = 600;
+	static GLFWwindow* _window;
+	//camera settings
+	Camera* _camera;
+	bool _firstMouse;
+	float _yaw;
+	float _pitch;
+	float _lastX;
+	float _lastY;
+	float _fov;
+
+	//engine clock time
+	static float _deltaTime;
+	static float _lastFrame;
+
+	//shader variables
+	const static unsigned int SHADOW_WIDTH = 1024;
```

Comment:
why are you making all of these variables static?

Reply:
It's pattern enforceable - I felt that while I started embedding pattern and way they are being used then after.

Reply:
what's the point of having consts static if they never change?

Reply:
and what do you mean by pattern enforceable?

Reply:
because you make an instance of an object?

Reply:
I will try to eliminate the consts bearing static if that wouldn't break anything.

Reply:
Let me give you a hint : https://www.reddit.com/r/cpp/comments/x7wxh/why_cant_a_derived_class_access_protected_members/. I am willing to give a lecture on template pattern and what's the difference while it came to applicability in C++ unlike language like JAVA. Do you agree ? I will add my confluence tomorrow or in coming week for your reference.

Reply:
> and what do you mean by pattern enforceable?

Let me give you a hint : https://www.reddit.com/r/cpp/comments/x7wxh/why_cant_a_derived_class_access_protected_members/. I am willing to give a lecture on template pattern and what's the difference when it comes to applicability in C++ unlike language like JAVA. Do you agree ? I will add my confluence tomorrow or in coming week for your reference.

Reply:
I need to conduct lecture for this thread! or Q&A I think.

Reply:
in what context does our code have a similar case than the example shown in that link?
```
class Base
{
protected:
  void DoSomething() {}
};

class Derived : public Base
{
public:
  void Test() 
  {
    Base base_class;
    base_class.DoSomething();
  }
};

int main(int argc, char* argv[])
{
  Derived der;
  der.Test();
  return 0;
}
```

Do we have a case where we have a function like "Test()" that creates a object of its own parent class?

if the base class's function is protected, couldn't "Test()" just do:
```
void Test() 
  {
    DoSomething();
  }
```

since it inherits the class already anyways?

Reply:
C++ uses some fairly odd scoped inheritance rules. When you have a protected member, you mean that instances deriving from the base are allowed free access to the field in this object. You are not saying that derivatives of the base are allowed to access any instance (which might not even be their own type). Consider:

class Base
{
protected:
    int x;
};

class Derived : public Base { };

class Sneaker : public Base
{
public:
    static void AlterX(Derived& d)
    {
        // One might think that I can access x here, because it comes from Base
        // and Sneaker inherits from Base; however, no regular casting will
        // allow me to do so (there is probably some incantation that does).
        d.x = 10;
    }
};

Reply:
> C++ uses some fairly odd scoped inheritance rules. When you have a protected member, you mean that instances deriving from the base are allowed free access to the field in this object. You are not saying that derivatives of the base are allowed to access any instance (which might not even be their own type). Consider:
> 
> class Base { protected: int x; };
> 
> class Derived : public Base { };
> 
> class Sneaker : public Base { public: static void AlterX(Derived& d) { // One might think that I can access x here, because it comes from Base // and Sneaker inherits from Base; however, no regular casting will // allow me to do so (there is probably some incantation that does). d.x = 10; } };

This was the case I was facing. You can't access protected 'x' in Sneaker class if you want directly.

Reply:
I see, that makes sense, but are you doing this type of thing anywhere?
In what case would a derived class need a reference to another derived class in our case?

Reply:
Nope, just try to convince yourself that protected member 'x' from the base class can't be accessed from the inherited class whatsoever. What they have portrayed is a way of accessing it - refer that comment inside the code Rose : )
Template is designed such that it needs direct access of protected data members of the Base class.

Reply:
> Nope, just try to convince yourself that protected member 'x' of the base class can't be accessed from the inherited class whatsoever. What they have portrayed is a way of accessing it - refer that comment inside the code Rose : ) Template is designed such that it needs direct access of protected data members of the Base class.

// One might think that I can access x here, because it comes from Base
// and Sneaker inherits from Base; however, no regular casting will
// allow me to do so (there is probably some incantation that does).

Reply:
That isn't true though. derived classes do get access to the protected variables of the base class.... The only reason why you aren't able to access them in this code you made is because you made all of the functions "const" and consts are implicitly static, which means they can only use static variables within them...

I don't understand why you made all those functions const in the first place

Reply:
Thanks for letting me know. I think I have been taking this documentation as a standard one : https://refactoring.guru/design-patterns/template-method/cpp/example. One way of doing this pattern.

Reply:
ahh ok, this makes a lot of sense. If this is the case then I guess everything is ok! :)

Reply:
If you feel the way I refactored is unsafe I can remove const and revised the code version as per.

Reply:
it should be good. i just needed more context, buit i understand now

Reply:
I am revisting c++  concepts once again :P

Reply:
me too haha